src/segmentmytif/features.py

```diff
@@ -27,42 +33,107 @@ def from_string(s):
-        loaded_features, _ = read_geotiff(features_path)
-        logger.info(f"Loading {feature_type.name} features (shape {loaded_features.shape}) from {features_path}")
-        return loaded_features
-
-
-def extract_features(input_data, feature_type, **extractor_kwargs):
+        return raster
+
+    if features_path is None:
+        features_path = get_features_path(raster_path, feature_type)
+    if not features_path.exists():
+        # Extract features
+        msg = (
+            f"No existing {feature_type.name} features found at {features_path} "
+            f"for input data with shape {raster.data.shape}"
+        )
+        logger.info(msg)
+
+        with log_duration(f"Extracting {feature_type.name} features", logger):
+            features_data = extract_features(raster.data, feature_type, chunk_overlap, **extractor_kwargs)
+        log_array(features_data, logger, array_name=f"{feature_type.name} features")
+
+        # Create xarray DataArray with the extracted features
+        # Keep the geospatial information from the input raster
+
+        features = raster.isel(band=0).drop_vars(["band"]).expand_dims(band=features_data.shape[0])
+        features.data = features_data
+
+        # Save the features to disk
+        msg = f"Saving {feature_type.name} features (shape {features_data.shape}) to {features_path}"
+        logger.info(msg)
+        features.rio.to_raster(features_path)
+
+    match compute_mode:
+        case "normal":
+            loaded_features = rioxarray.open_rasterio(features_path)
+        case "parallel" | "safe":
+            # Unpack the chunk sizes from the input raster
+            chunks = {
+                "band": raster.chunksizes["band"][0],
+                "x": raster.chunksizes["x"][0],
+                "y": raster.chunksizes["y"][0],
+            }
+            loaded_features = rioxarray.open_rasterio(features_path, chunks=chunks)
+    msg = f"Loading {feature_type.name} features (shape {loaded_features.shape}) from {features_path}"
+    logger.info(msg)
+
+    return loaded_features
+
+
+def extract_features(input_data, feature_type, chunk_overlap=16, **extractor_kwargs):
     extractor = {
-        FeatureType.IDENTITY: extract_identity_features,
         FeatureType.FLAIR: extract_flair_features,
     }[feature_type]
 
-    return extractor(input_data, **extractor_kwargs)
+    # If dask array, map feature extraction function to each block
+    # The blockwise extraction is only applied to FLAIR features
+    # FeatureType.IDENTITY directly returns the input data, thus no need to map_overlap
+    if isinstance(input_data, Array):
+        # Make template dask array according to the extractor
+        if feature_type == FeatureType.FLAIR:
+            # If FLAIR, output features has bands * NUM_FLAIR_CLASSES
+            meta = da.zeros_like(
+                input_data,
+                shape=(input_data.shape[0] * NUM_FLAIR_CLASSES, input_data.shape[1], input_data.shape[2]),
```

Comment:
It is a bit of a bummer that there's now flair-specific code here. I had hoped to keep code as this level general, and move all flair related code to another file later on. This would especially help keep things tidy if we would add more feature types in the future. However, it's not the most important thing and maybe there won't be any new feature types ever. You decide if it's worth it to come up with a solution code-wise to push that flair specific logic down ;-)